src/ray/util/util_test.cc

```diff
@@ -0,0 +1,129 @@
+  ASSERT_EQ(ParseCommandLine(R"(\\a)", posix), ArgList({R"(\a)"}));
+  ASSERT_EQ(ParseCommandLine(R"(\\a)", win32), ArgList({R"(\\a)"}));
+  ASSERT_EQ(ParseCommandLine(R"(\\\a)", posix), ArgList({R"(\a)"}));
+  ASSERT_EQ(ParseCommandLine(R"(\\\a)", win32), ArgList({R"(\\\a)"}));
+  ASSERT_EQ(ParseCommandLine(R"(\\)", posix), ArgList({R"(\)"}));
+  ASSERT_EQ(ParseCommandLine(R"(\\)", win32), ArgList({R"(\\)"}));
+  ASSERT_EQ(ParseCommandLine(R"("\\a")", posix), ArgList({R"(\a)"}));
+  ASSERT_EQ(ParseCommandLine(R"("\\a")", win32), ArgList({R"(\\a)"}));
+  ASSERT_EQ(ParseCommandLine(R"("\\\a")", posix), ArgList({R"(\\a)"}));
+  ASSERT_EQ(ParseCommandLine(R"("\\\a")", win32), ArgList({R"(\\\a)"}));
+  ASSERT_EQ(ParseCommandLine(R"('a'' b')", posix), ArgList({R"(a b)"}));
+  ASSERT_EQ(ParseCommandLine(R"('a'' b')", win32), ArgList({R"('a'')", R"(b')"}));
+  ASSERT_EQ(ParseCommandLine(R"('a')", posix), ArgList({R"(a)"}));
+  ASSERT_EQ(ParseCommandLine(R"('a')", win32), ArgList({R"('a')"}));
+  ASSERT_EQ(ParseCommandLine(R"(x' a \b')", posix), ArgList({R"(x a \b)"}));
+  ASSERT_EQ(ParseCommandLine(R"(x' a \b')", win32), ArgList({R"(x')", R"(a)", R"(\b')"}));
+}
+
+TEST(UtilTest, CreateCommandLineTest) {
+  typedef std::vector<std::string> ArgList;
+  CommandLineSyntax posix = CommandLineSyntax::POSIX, win32 = CommandLineSyntax::Windows,
+                    all[] = {posix, win32};
+  std::vector<ArgList> test_cases({
+      ArgList({R"(a)"}),
+      ArgList({R"(a b)"}),
+      ArgList({R"(")"}),
+      ArgList({R"(')"}),
+      ArgList({R"(\)"}),
+      ArgList({R"(/)"}),
+      ArgList({R"(#)"}),
+      ArgList({R"($)"}),
+      ArgList({R"(!)"}),
+      ArgList({R"(@)"}),
+      ArgList({R"(`)"}),
+      ArgList({R"(^)"}),
+      ArgList({R"(&)"}),
+      ArgList({R"(|)"}),
+      ArgList({R"(a")", R"('x)", R"(?^'"{)", R"(]))", R"(!)", R"(~`\)"}),
+  });
+  for (CommandLineSyntax syn : all) {
+    for (const ArgList &arglist : test_cases) {
+      ASSERT_EQ(ParseCommandLine(CreateCommandLine(arglist, syn), syn), arglist);
+      std::string cmdline = CreateCommandLine(arglist, syn);
+      std::string buf((2 + cmdline.size()) * 6, '\0');
+      std::string test_command = std::string(argv0);
+      test_command = "\"" + test_command + "\"";
+#ifdef _WIN32
+      test_command = "\"" + test_command;
+#endif
+      test_command += " --println " + cmdline;
+#ifdef _WIN32
+      test_command = test_command + "\"";
+#endif
+      FILE *proc;
+#ifdef _WIN32
+      proc = syn == win32 ? _popen(test_command.c_str(), "r") : NULL;
+#else
+      proc = syn == posix ? popen(test_command.c_str(), "r") : NULL;
+#endif
+      if (proc) {
+        ASSERT_TRUE(fgets(&*buf.begin(), static_cast<int>(buf.size()), proc));
+        buf = std::string(buf.c_str());  // remove trailing nulls
+        ASSERT_TRUE(!buf.empty());
+        ASSERT_EQ(buf.back(), '\n');
+        buf.pop_back();
+        ASSERT_EQ(ParseCommandLine(cmdline, syn), arglist);
```

Comment:
We already tested this above in `ASSERT_EQ(ParseCommandLine(CreateCommandLine(arglist, syn), syn), arglist);`, didn't we? Is this line supposed to test something involving `buf`?

Reply:
Oops, yeah! Good eye, I'll fix it. Hopefully this hasn't let bugs slip through!